Allocate memory for fake_fcinfo

jeffcoat opened the pull request:
Only applies to PG12+.

I can reproduce the crash from https://github.com/plv8/plv8/issues/373 and https://github.com/plv8/plv8/issues/370. This patch fixes it.

More precisely, the change to plv8.cc fixed all the problems that I can reproduce; I made the same change to plv8_func.cc because they're so structurally similar.

mjeffcoat-foreflight committed:
Allocate memory for fake_fcinfo before memsetting

Only applies to PG12+.

JerrySievert commented:
awesome! I'll wait to see what CI looks like before merging.

jeffcoat commented:
I'm very new to this code and PG extensions and don't really know what I'm doing. I'm mostly hoping this will point someone who does in the right direction.

JerrySievert commented:
hm. reviewing this, it doesn't actually do what you might think it does, and also causes a memory leak.

I'll comment in-code.

JerrySievert commented on plv8.cc:
```diff
@@ -1620,7 +1620,8 @@ GetGlobalContext(Persistent<Context>& global_context)
 				Datum ret = has_function_privilege_id(&fake_fcinfo);
 #else
-				MemSet(fake_fcinfo, 0, sizeof(fake_fcinfo));
+				fake_fcinfo = (FunctionCallInfo) palloc(SizeForFunctionCallInfo(2));
```
here you're allocating over an existing object.

JerrySievert replied:
also missing a `pfree()` so causing a memory leak

jeffcoat replied:
I can believe that. (Though I don't see how. There's a declaration `FunctionCallInfo fake_fcinfo;` on line 1601, then this is the first statement that refers to fake_fcinfo following the declaration.)

Without the allocation, I get a segfault at the line `MemSet(fake_fcinfo, 0, sizeof(fake_fcinfo));` I'm interpreting that as an attempt to zero out unallocated memory.

Backtrace from a core file:

```
Core was generated by `postgres: 12/main: import import ::1(42918) SELECT                            '.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007fdeeb657a2c in GetGlobalContext (global_context=...) at plv8.cc:1625
1625					fake_fcinfo->flinfo = &flinfo;
[Current thread is 1 (Thread 0x7fdf710b8a40 (LWP 1747400))]
(gdb) bt
#0  0x00007fdeeb657a2c in GetGlobalContext (global_context=...) at plv8.cc:1625
```

Which shows in syslog as
```
Dec 16 17:01:39 behenian kernel: [1559398.124502] postgres[1747400]: segfault at 0 ip 00007fdeeb657a2c sp 00007ffc33f6d320 error 6 in plv8-3.0alpha.so[7fdeeb64a000+934000]
Dec 16 17:01:39 behenian kernel: [1559398.124517] Code: 1f 80 00 00 00 00 48 83 c0 08 48 c7 40 f8 00 00 00 00 48 39 d0 75 ef 48 8b 05 c8 e3 b9 00 31 ff c7 85 d8 fe ff ff 00 00 00 00 <48> 89 0f 48 8b 00 48 89 85 f0 fe ff ff b8 02 00 00 00 66 89 47 1e
```

JerrySievert replied:
likely what occurred is that Postgres internals changed the `typedef` for FunctionCallInfo.  I'll take a deeper look into it when I get the time.

JerrySievert replied:
if that's the case, then it will need to be `#if`'d to work correctly across all postgres versions that plv8 supports.

jeffcoat replied:
Thank you for your feedback. 

* You're right about the changed typedef for FunctionCallInfo. You set up the  `#if`s a few months ago in commit 77b941128cdb; I'm building off that.

* To avoid memory leaks, I've changed fake_fcinfo to be stack-allocated, instead of using palloc(). The LOCAL_FCINFO macro is from the PG12 version of fmgr.h.

mjeffcoat-foreflight committed:
Allocate FunctionCallInfo on the stack.

Replace palloc() with a macro from fmgr.h
to put fake_fcinfo on the stack, so we
don't have to worry about pfree().

jeffcoat commented:
I don't think the CI build failure was caused by a problem on this branch, but I don't have the authority to re-try it.

JerrySievert commented:
I restarted the job

jeffcoat commented:
Thank you; much better.

mjeffcoat-foreflight committed:
Specify extra_float_digits for windows test.

Related to Postgres commit 02ddd499322ab6f2,
which changed the default value from 0 to 1.

Explicitly setting the value here means the
output will be formatted in the same way
across versions.

mjeffcoat-foreflight committed:
PG12 version of plv8.out.

The only difference is an error message changing
from
  invalid input syntax for integer:
to
  invalid input syntax for type integer:

mjeffcoat-foreflight committed:
Add PGVERSION=12 to env matrix.

JerrySievert commented on expected/plv8_2.out:
```diff
@@ -0,0 +1,740 @@
+-- CREATE FUNCTION
```
what are you trying to add here?

jeffcoat replied:
Sorry; it's explained in the commit message, but this UI doesn't show that unless you go looking for it.

One of the tests in plv8.sql -- `text_to_int4('abc')` -- fails under PG12 because the error message changed from `invalid input syntax for integer:` to `invalid input syntax for type integer:`.

I was trying to make the tests pass on all versions with minimal re-writing, so I added plv8_2.out as an alternative expected/ file that uses the new error message. I'd be happy to change it if you'd prefer split this case into a separate test file, or something like that.

JerrySievert replied:
I think I'd prefer a second PR, separate out the concerns.

jeffcoat replied:
No problem; I put them together here because the tests won't pass against PG12 without the fcinfo change. Which speaks well for the test suite.

JerrySievert replied:
updating `.travis.yml` should also go in the other PR :)
